pkg/operator/sync.go

```diff
@@ -830,7 +831,23 @@ func (optr *Operator) getGlobalConfig() (*configv1.Infrastructure, *configv1.Net
 	if err != nil && !apierrors.IsNotFound(err) {
 		return nil, nil, nil, nil, err
 	}
-	return infra, network, proxy, dns, nil
+	return infra, network, removeEtcdNoProxyConfig(proxy), dns, nil
+}
+
+// removeEtcdNoProxyConfig temporary removal of etcd NoProxy records
+// this allows MCD to tolerate divergence of on disk state until records
+// are removed from installer.
+func removeEtcdNoProxyConfig(proxy *configv1.Proxy) *configv1.Proxy {
+	if proxy != nil && proxy.Status != (configv1.ProxyStatus{}) {
+		filtered := []string{}
+		for _, entry := range strings.Split(proxy.Status.NoProxy, ",") {
+			if !regexp.MustCompile(`^etcd-.*$`).MatchString(entry) {
+				filtered = append(filtered, entry)
+			}
+			proxy.Status.NoProxy = strings.Join(filtered, ",")
```

Comment:
The way filtered is built up and then incrementally assigned to the field of the input argument is pretty unusual to me, at first I thought I was seeing a bug. To remove the side effects entirely and make it even more clear, what do you think about moving the precondition check to the caller, like:

```go
func (optr *Operator) getGlobalConfig() (*configv1.Infrastructure, *configv1.Net) {
	// <omitted>
  
	if proxy != nil && proxy.Status != (configv1.ProxyStatus{}) {
		proxy.Status.NoProxy = removeEtcdNoProxyConfig(proxy.Status.NoProxy)
	}
	return infra, network, proxy, dns, nil
}

func removeEtcdNoProxyConfig(lines string) string {
	filtered := []string{}
	for _, entry := range strings.Split(lines, ",") {
		if !regexp.MustCompile(`^etcd-.*$`).MatchString(entry) {
			filtered = append(filtered, entry)
		}
	}
	return strings.Join(filtered, ",")
}
```